### PR TITLE
[move-lang] do not coalesce nested loops into a shared header

### PR DESCRIPTION
**Pull request by mengxu-fb**
In the source compiler, there is a simple optimization performed which removes any "fall through jumps" --- a direct jump to the next instruction.

However, if a fall-through jump is *the only instruction* in a loop header block, the jump should be preserved, because removing this jump means removing the header block as well, which essentially coalesced the outer loop to share the same header block with the inner loop. This is absolutely correct for program semantics, but creates a hard time for loop invariant instrumentation.

To illustrate, given a similar move program
```
// loop A
loop {
  // loop B
  loop {
    // loop C
    loop { 
      ... 
    } 
    ...
  }
...
}
```
where `loop` is the first statement in a loop body, then all loops will be coalesced to using the same header. In the above case, loop A, B, C will share the same header block in the control-flow graph.

But for the loop invariant instrumentation, we need these loops to have different headers such that we can instrument invariants for each loop respectively.

To illustrate this, here is the Move code

```
module VerifyLoopsWithMemoryOps {
    use 0x1::Vector;

    spec module {
        pragma verify=true;
    }

    public fun nested_loop(a: &mut vector<u64>, b: &mut vector<u64>) {
        let length = Vector::length(a);
        spec {
            assume length > 0;
            assume length == len(b);
        };
        let i = 0;
        let x = Vector::borrow_mut(a, i);
        let y = Vector::borrow_mut(b, i);
        loop {
            loop {
                loop {
                    spec {
                        assert x != y;
                    };
                    if (*x <= *y) {
                        break
                    };
                    *y = *y + 1;
                };

                if (*y <= *x) {
                    break
                };
                *x = *x + 1;
            };
            if (i == length) {
                break
            };
            i = i + 1;
            x = Vector::borrow_mut(a, i);
            y = Vector::borrow_mut(b, i);
        };
        spec {
            assert forall m in 0..len(a): a[m] == b[m];
        };
    }
    spec fun nested_loop {
        aborts_if false;
    }
}
```

Here is the
[disassembled bytecode w/o coalescing](https://github.com/diem/diem/files/6208349/loops-after.txt), and
[CFG w/o coalescing](https://github.com/diem/diem/files/6208356/cfg.pdf)
which can be compared with the coalesced version:
[disassembled bytecode w coalescing](https://github.com/diem/diem/files/6208467/loop.txt), and
[CFG w/ coalescing](https://github.com/diem/diem/files/6208446/cfg.pdf)
Pay attention to the loop header blocks that have only a single GOTO statement.

## Motivation

Make loop invariant instrumentation easier

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

This should not break anything, so the CI should all pass.

To reproduce the example:
`mvp --dependency <path-to-move-stdlib-modules>/Vector.move --dump-bytecode --dump-cfg loops.move`
`dot -Tpdf loops_10_verification_analysis_VerifyLoopsWithMemoryOps__nested_loop_baseline_cfg.dot -o cfg.pdf`

### CODE REVIEW

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```

File: language/move-lang/src/shared/graph.rs [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+            (&natural_loops[1], &natural_loops[0])
+        };
+
+        assert_eq!(inner_loop.loop_header, 2);
+        assert_eq!(inner_loop.loop_body, vec![2, 3].into_iter().collect());
+
+        assert_eq!(outer_loop.loop_header, 2);
+        assert_eq!(outer_loop.loop_body, vec![2, 3, 4, 5].into_iter().collect());
+    }
+}
```
Comment (mengxu-fb) [inline language/move-lang/src/shared/graph.rs:360]: This file is copied from #7952 which originally exists as `move-prover/bytecode/src/graph.rs`

**Review by tnowacki — CHANGES_REQUESTED** [review]
Let's chat about this, if I understand the problem correctly, this can be done MUCH MUCH easier

**Comment by mengxu-fb** [comment]
Closing in favor of #8049.

**Pull request closed without merging**